### PR TITLE
Global object escape routine.

### PR DESCRIPTION
**Pull request by zherczeg**
JerryScript-DCO-1.0-Signed-off-by: Zoltan Herczeg zherczeg@inf.u-szeged.hu

### CODE REVIEW

**Comment by zherczeg** [comment]
I know a guard is missing, but otherwise the patch is ready for review.

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 1]
```diff
@@ -1164,6 +1164,138 @@ ecma_builtin_global_object_encode_uri (ecma_value_t this_arg __attr_unused___, /
...
+
+  lit_utf8_byte_t *output_char_p = output_start_p;
+
+  lit_utf8_iterator_seek_bos (&iterator);
+
+  while (!lit_utf8_iterator_is_eos (&iterator))
+  {
+    ecma_char_t chr = lit_utf8_iterator_read_next (&iterator);
+
+    if (chr < 128)
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: @zherczeg, could you, please clarify origin of the constant?

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 2]
```diff
@@ -1164,6 +1164,138 @@ ecma_builtin_global_object_encode_uri (ecma_value_t this_arg __attr_unused___, /
...
+    {
+      if (ecma_builtin_global_object_character_is_in ((uint32_t) chr, escape_set))
+      {
+        *output_char_p = (lit_utf8_byte_t) chr;
+        output_char_p++;
+      }
+      else
+      {
+        ecma_builtin_global_object_byte_to_hex (output_char_p, (lit_utf8_byte_t) chr);
+        output_char_p += 3;
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: `URI_ENCODED_BYTE_SIZE`?

**Comment by ruben-ayrapetyan** [comment]
@zherczeg, could you, please, replace number values with corresponding named constants (if they are local) / preprocessor definitions (if they are global), and add `ecma_` (maybe `ecma_uri_`) / `ECMA_` prefixes to global variables?

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 3]
```diff
@@ -1164,6 +1164,138 @@ ecma_builtin_global_object_encode_uri (ecma_value_t this_arg __attr_unused___, /
...
 } /* ecma_builtin_global_object_encode_uri */
 
+/*
+ * Escape characters bitset:
+ *   One bit for each character between 0 - 127.
+ *   Bit is set if the character does not need to be converted to %xx form.
+ */
+static const uint8_t escape_set[16] =
+{
+  0x0, 0x0, 0x0, 0x0, 0x0, 0xec, 0xff, 0x3,
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: Could you, please, clarify how ECMA-262 defined the values, i.e. how can the values be calculated?

**Comment by zherczeg** [comment]
I fixed your requests.

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 4]
```diff
@@ -1164,6 +1164,149 @@ ecma_builtin_global_object_encode_uri (ecma_value_t this_arg __attr_unused___, /
...
+ */
+#define ESCAPE_ENCODED_UNICODE_CHARACTER_SIZE (6)
+
+/*
+ * Escape characters bitset:
+ *   One bit for each character between 0 - 127.
+ *   Bit is set if the character does not need to be converted to %xx form.
+ *   These characters are: a-z A-Z 0-9 @ * _ + - . /
+ */
+static const uint8_t escape_set[16] =
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: `escape_set`

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 5]
```diff
@@ -1164,6 +1164,149 @@ ecma_builtin_global_object_encode_uri (ecma_value_t this_arg __attr_unused___, /
...
 
+/*
+ * Maximum value of a byte.
+ */
+#define ESCAPE_MAXIMUM_BYTE_VALUE (255)
+
+/*
+ * Format is a percent sign followed by lowercase u and four hex digits.
+ */
+#define ESCAPE_ENCODED_UNICODE_CHARACTER_SIZE (6)
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: `ECMA_ESCAPE_ENCODED_UNICODE_CHARACTER_SIZE`

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 6]
```diff
@@ -1164,6 +1164,149 @@ ecma_builtin_global_object_encode_uri (ecma_value_t this_arg __attr_unused___, /
   return ecma_builtin_global_object_encode_uri_helper (uri, unescaped_uri_set);
 } /* ecma_builtin_global_object_encode_uri */
 
+/*
+ * Maximum value of a byte.
+ */
+#define ESCAPE_MAXIMUM_BYTE_VALUE (255)
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: `ECMA_ESCAPE_MAXIMUM_BYTE_VALUE`

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 7]
```diff
@@ -1164,6 +1164,149 @@ ecma_builtin_global_object_encode_uri (ecma_value_t this_arg __attr_unused___, /
...
+
+/*
+ * Escape characters bitset:
+ *   One bit for each character between 0 - 127.
+ *   Bit is set if the character does not need to be converted to %xx form.
+ *   These characters are: a-z A-Z 0-9 @ * _ + - . /
+ */
+static const uint8_t escape_set[16] =
+{
+  0x0, 0x0, 0x0, 0x0, 0x0, 0xec, 0xff, 0x3,
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: Where did you get this constants? RFC 3986?

How does this list corresponds to the following from spec:
```
uriAlpha ::: one of
a b c d e f g h i j k l m n o p q r s t u v w x y z
A B C D E F G H I J K L M N O P Q R S T U V W X Y Z
uriMark ::: one of
- _ . ! ~ * ' ( )
```

Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: http://www.ecma-international.org/ecma-262/5.1/#sec-B.2.1

**Comment by zherczeg** [comment]
Any more comments?

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 8]
```diff
@@ -1026,7 +1026,7 @@ ecma_builtin_global_object_byte_to_hex (lit_utf8_byte_t *dest_p, /**< destinatio
  */
 static ecma_completion_value_t
 ecma_builtin_global_object_encode_uri_helper (ecma_value_t uri, /**< uri argument */
-                                              uint8_t* unescaped_uri_bitset) /**< unescaped bitset */
+                                              const uint8_t *unescaped_uri_bitset) /**< unescaped bitset */
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: `_p`

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 9]
```diff
@@ -1180,6 +1180,153 @@ ecma_builtin_global_object_encode_uri_component (ecma_value_t this_arg __attr_un
...
+    }
+    else if (chr > ECMA_ESCAPE_MAXIMUM_BYTE_VALUE)
+    {
+      /*
+       * Although ecma_builtin_global_object_byte_to_hex inserts a percent (%) sign
+       * the follow-up changes overwrites it. We call this function twice to
+       * produce four hexadecimal characters (%uxxxx format).
+       */
+      ecma_builtin_global_object_byte_to_hex (output_char_p + 3, (lit_utf8_byte_t) (chr & 0xff));
+      ecma_builtin_global_object_byte_to_hex (output_char_p + 1, (lit_utf8_byte_t) (chr >> 8));
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: We have `JERRY_BITSINBYTE` for 8 in jrt.h

**Comment by egavrin** [comment]
Except minor things that I missed last time - good to me.

**Comment by zherczeg** [comment]
Thank you, fixed.

**Comment by LaszloLango** [comment]
+1 lgtm

**Comment by zherczeg** [comment]
Landed in bf780ad3dc6fe336d6741a8cdc9d6d4f51f3e95e

**Pull request closed without merging**